startups/lib/main_9.dart

```diff
@@ -0,0 +1,220 @@
+        page = GeneratorPage();
+        break;
+      case 1:
+        page = FavoritesPage();
+        break;
+      default:
+        throw UnimplementedError();
+    }
+
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return Scaffold(
+          body: Row(
+            children: [
+              SafeArea(
+                child: NavigationRail(
+                  extended: constraints.maxWidth >= 600,
+                  destinations: [
+                    NavigationRailDestination(
+                      icon: Icon(Icons.home),
+                      label: Text('Home'),
+                    ),
+                    NavigationRailDestination(
+                      icon: Icon(Icons.favorite),
+                      label: Text('Favorites'),
+                    ),
+                  ],
+                  selectedIndex: selectedIndex,
+                  onDestinationSelected: (int value) => setState(() {
+                    selectedIndex = value;
+                  }),
+                ),
+              ),
+              Expanded(
+                child: Container(
+                  color: Theme.of(context).colorScheme.secondaryContainer,
+                  child: page,
+                ),
+              ),
+            ],
+          ),
+        );
+      },
+    );
+  }
+}
+
+class GeneratorPage extends StatelessWidget {
+  @override
+  Widget build(BuildContext context) {
+    var appState = context.watch<MyAppState>();
+
+    return Column(
+      crossAxisAlignment: CrossAxisAlignment.stretch,
+      children: [
+        Expanded(
+          flex: 2,
+          child: ListView(
+            reverse: true,
+            children: [
+              for (var historicalPair in appState.history)
+                Center(
+                  child: Text(historicalPair.asLowerCase),
+                ),
+            ],
+          ),
+        ),
+        SizedBox(height: 16),
+        BigCard(appState.current),
+        Row(
+          mainAxisAlignment: MainAxisAlignment.end,
+          children: [
+            ElevatedButton.icon(
+              icon: Icon(appState.favorites.contains(appState.current)
+                  ? Icons.favorite
+                  : Icons.favorite_border),
+              label: Text('Like'),
+              onPressed: () {
+                appState.toggleCurrentFavorite();
+              },
+            ),
+            SizedBox(width: 8),
+            ElevatedButton.icon(
+              icon: Icon(Icons.refresh),
+              label: Text('Next'),
+              onPressed: () {
+                appState.generateNext();
+              },
+            ),
+            SizedBox(width: 28),
+          ],
+        ),
+        Spacer(),
+      ],
+    );
+  }
+}
+
+class BigCard extends StatelessWidget {
+  final WordPair pair;
+
+  BigCard(this.pair);
+
+  @override
+  Widget build(BuildContext context) {
+    var theme = Theme.of(context);
+    var style = theme.textTheme.displayMedium!.copyWith(
```

Comment:
The `!` here is unfortunate, but at least it gives us a reason to talk about null safety.